daemons/pacemakerd/pacemakerd.c

```diff
@@ -33,31 +33,12 @@
 #include <ctype.h>
 
 static gboolean fatal_error = FALSE;
-static GMainLoop *mainloop = NULL;
 static bool global_keep_tracking = false;
 
 #define PCMK_PROCESS_CHECK_INTERVAL 5
 
-static crm_trigger_t *shutdown_trigger = NULL;
-static crm_trigger_t *startup_trigger = NULL;
 static const char *pid_file = PCMK_RUN_DIR "/pacemaker.pid";
 
-/* state we report when asked via pacemakerd-api status-ping */
-static const char *pacemakerd_state = XML_PING_ATTR_PACEMAKERDSTATE_INIT;
-static gboolean running_with_sbd = FALSE; /* local copy */
-/* When contacted via pacemakerd-api by a client having sbd in
- * the name we assume it is sbd-daemon which wants to know
- * if pacemakerd shutdown gracefully.
- * Thus when everything is shutdown properly pacemakerd
- * waits till it has reported the graceful completion of
- * shutdown to sbd and just when sbd-client closes the
- * connection we can assume that the report has arrived
- * properly so that pacemakerd can finally exit.
- * Following two variables are used to track that handshake.
- */
-static unsigned int shutdown_complete_state_reported_to = 0;
-static gboolean shutdown_complete_state_reported_client_closed = FALSE;
-
```

Comment:
Some of these are debatable since the main file should stay in control of the main loop and the flow of the main process itself, but we can leave that for the longer term -- it may be worth isolating some of this with some accessor functions.

Reply:
All this stuff (and the pcmk_shutdown calls) seem pretty tough to split out.  They're embedded pretty deeply in these functions, all of which are themselves pretty tangled up in calling each other and being passed around as callbacks.